src/ZEO/asyncio/_futures.pyx

```diff
@@ -0,0 +1,489 @@
+# if fut does not support cancelling with message, the message is saved in fut._xasyncio_cancel_msg .
+cdef _cancel_future(fut, msg):
+    try:
+        return fut.cancel(msg)
+    except TypeError:
+        # on trollius and py3 < 3.9 Future.cancel does not accept msg
+        _ = fut.cancel()
+        fut._xasyncio_cancel_msg = msg
+        return _
+
+
+cdef class AsyncTask(Future):
+    """Simplified ``asyncio.Task``.
+
+    Steps are not scheduled but executed immediately.
+    """
+    cdef CoroutineExecutor executor
+
+    def __init__(self, coro, loop=None):
+        Future.__init__(self, loop=loop)
+        self.executor = CoroutineExecutor(self, coro)  # reference cycle
+        self.executor.step()
+
+    cpdef cancel(self, msg=None):
+        """external cancel request."""
+        return self.executor.cancel(msg)
+
+    def _cancel(self, msg):
+        """internal cancel request."""
+        return Future.cancel(self, msg)
+
+
+cdef class ConcurrentTask(ConcurrentFuture):
+    """Task reporting to ``ConcurrentFuture``.
+
+    Steps are not scheduled but executed immediately.
+    Cancel can be used from any thread.
+    """
+    cdef CoroutineExecutor executor
+    cdef long loop_thread_id
+
+    def __init__(self, coro, loop):
+        ConcurrentFuture.__init__(self, loop=loop)
+        self.executor = CoroutineExecutor(self, coro)  # reference cycle
+        self.loop_thread_id = -1
+        loop.call_soon_threadsafe(self._start)
+
+    cpdef _start(self):
+        # asyncio.Loop has ._thread_id, but uvloop does not expose it
+        self.loop_thread_id = get_ident()  # with gil
+        self.executor.step()
+
+    cpdef cancel(self, msg=None):
+        """external cancel request.
+
+        cancel requests cancellation of the task.
+        it is safe to call cancel from any thread.
+        """
+        # invoke CoroutineExecutor.cancel on the loop thread and wait for its result.
+        # but run it directly to avoid deadlock if we are already on the loop thread.
+        if get_ident() == self.loop_thread_id:  # with gil
+            return self.executor.cancel(msg)
+        self._cancel_via_loopthread(msg)
+
+    def _cancel_via_loopthread(self, msg):  # cpdef does not allow to use closures
+        sema = Lock()
+        sema.acquire()
+        res = [None]
+        def _():
+            try:
+                x = self.executor.cancel(msg)
+            except BaseException as e:
+                x = e
+                if PY_MAJOR_VERSION < 3: # trollius stops the loop by raising _StopError
+                    if isinstance(e, asyncio.base_events._StopError):
+                        x = True
+                        raise
+            finally:
+                res[0] = x
+                sema.release()
+        self._loop.call_soon_threadsafe(_)
+        sema.acquire() # wait for the call to complete
+        r = res[0]
+        if isinstance(r, BaseException):
+            raise r
+        return r
```

Comment:
Thanks for the note. Here is that `_chain_future`:

https://github.com/python/cpython/blob/61762b93871419b34f02d83cee5ca0d94d4a2903/Lib/asyncio/futures.py#L365-L406

what it basically does regarding cancellation is that it attaches `loop.call_soon_threadsafe(fut2.cancel)` to `fut1` done callback if fut1 is cancelled.

Maybe I'm missing something but I'm not sure how it helps us here. Here we need to initiate such cancellation, not to propagate a cancellation from another future. Initiating cancellation is easy - just do `loop.call_soon_threadsafe(executor.cancel)` - and it is already part of that code (see line 450). But we also want to wait for the result of that cancel and to return that result to the original caller. That's why semaphore is added for such wait/wakeup. Yes, `call_soon_threadsafe` can be used from any thread, but if run originally on the loop thread, initiate the semaphore, invoke loop.call_soon_threadsafe and then initiate waiting on `sema.acquire`, the loop thread will block without having a chance to execute what was scheduled. In other words it will turn out as deadlock even if calling just `call_soon_threadsafe` is ok from any thread.

Once again maybe I'm missing something. If it is indeed the case could you please suggest how you would implement this part in a simpler way?

P.S. one thing I noticed now is the missing `return` on L432. Though py version does not have that problem:

https://github.com/zopefoundation/ZEO/blob/11357a8366a46a2b5525c47a0e7f5c26bbfc336d/src/ZEO/asyncio/futures.py#L414-L445

Reply:
> P.S. one thing I noticed now is the missing return on L432.

Should be hopefully fixed by https://github.com/zopefoundation/ZEO/commit/a8f395e3 .